From 3beb5342423e8f4358395d15d48bdbf3a2dfbfee Mon Sep 17 00:00:00 2001
From: Serge Hallyn <serge.hallyn@ubuntu.com>
Date: Fri, 8 Jan 2016 12:20:01 -0800
Subject: [PATCH 2/2] simplify and fix must_strcat_pid

Like the last commit by Wolfgang, merge the alloc cases into one
realloc.  Dereference *src as it must be after all.

Signed-off-by: Serge Hallyn <serge.hallyn@ubuntu.com>
---
 lxcfs.c | 20 +++++++-------------
 1 file changed, 7 insertions(+), 13 deletions(-)

diff --git a/lxcfs.c b/lxcfs.c
index abcbf3a..9859171 100644
--- a/lxcfs.c
+++ b/lxcfs.c
@@ -74,27 +74,21 @@ struct file_info {
  */
 static void must_strcat_pid(char **src, size_t *sz, size_t *asz, pid_t pid)
 {
-	char *d = *src;
 	char tmp[30];
 
 	int tmplen = sprintf(tmp, "%d\n", (int)pid);
 
-	if (!d) {
+	if (!*src || tmplen + *sz + 1 >= *asz) {
+		char *tmp;
 		do {
-			d = malloc(BUF_RESERVE_SIZE);
-		} while (!d);
-		*src = d;
-		*asz = BUF_RESERVE_SIZE;
-	} else if (tmplen + *sz + 1 >= *asz) {
-		do {
-			d = realloc(src, *asz + BUF_RESERVE_SIZE);
-		} while (!d);
-		*src = d;
+			tmp = realloc(*src, *asz + BUF_RESERVE_SIZE);
+		} while (!tmp);
+		*src = tmp;
 		*asz += BUF_RESERVE_SIZE;
 	}
-	memcpy(d+*sz, tmp, tmplen);
+	memcpy((*src) +*sz , tmp, tmplen);
 	*sz += tmplen;
-	d[*sz] = '\0';
+	(*src)[*sz] = '\0';
 }
 
 static int wait_for_pid(pid_t pid)
-- 
2.5.0

